### ISSUE TITLE
Document how to integrate well

### ISSUE THREAD

**Issue by ethanfrey** [CosmWasm/token-factory#8]
We need some changes to app.go.

Registering the custom callbacks is easy, just note the addition of the supportedFeature.
Also that x/tokenfactory module must also be added (like 10 places).

I think most changes are in #7 but it would be good to have a short file explaining how to add this to an existing chain (also how to configure the params)

**Comment by jhernandezb** [comment]
Stargaze uses `WithMessageEncoders`  for some custom bindings but the docs says it should not be combined with `WithMessageHandlerDecorator` or `WithMessageHandler`, but I still fail to understand the difference between these 3.

**Comment by ethanfrey** [comment]
TL;DR: If you mess up, app.go will panic on constructing the app. If it doesn't panic, it should be fine.

WithMessageHandler replaces the existing one, and [WithMessageHandlerDecorator  wraps whatever is there](https://github.com/CosmWasm/wasmd/blob/54fec05c1198fbbf58d123ae30a2e622d3997b1c/x/wasm/keeper/options.go#L27-L41)

[WithMessageEncoders](https://github.com/CosmWasm/wasmd/blob/54fec05c1198fbbf58d123ae30a2e622d3997b1c/x/wasm/keeper/options.go#L75-L82) makes a number of assumptions on what is stored in `k.messenger` and panics if it was replaced before.

Thus, you *should* be able to do `k.WithMessageEncoders().WithMessageHandlerDecorator()` safely. But not the other combinations.

**Comment by ethanfrey** [comment]
However, this brings up a good point - how to combine with other custom messages.

This one claims the `token` namespace and puts all messages/queries there to avoid grabbing a lot of top-level keys and conflicting with other people's custom messages (Mint is a common name).

You will need to figure out how to combine this with yours, and may need to forgo `RegisterCustomPlugins` use the lower level items. I guess we may also need to handle custom messages that don't use the token field (existing Stargaze messages) differently: https://github.com/CosmWasm/token-factory/blob/1345f322c8877322d974c316d46c0b0aa6f24e80/x/tokenfactory/bindings/message_plugin.go#L46-L48

I don't really know your codebase, but hopefully some hints to get this working. If you can add this in a way that app.go doesn't panic and existing tests don't break, it is a good chance it works. As mentioned above, you may want to make a PR to change what we do when we don't find a `token` field when deserialising CustomMsg.

**Comment by jhernandezb** [comment]
Thank you will take a look but I figure as CosmWasm it's getting more mainstream integrations we will see multiple modules with custom binding and a namespace collisions, we just have 2 messages so I don't find it difficult to integrate (I hope so 😬 )

**Comment by ethanfrey** [comment]
If you see how I wrap all messages in one top level namespace (Token), I want to promote this as a pattern.

Maybe another one for `OsmosisSwap` with all such messages. And maybe some other features shared between multiple chains, like, say `ICA`

Grouping to allow freedom within one namespace. This repo should also show how people outside of Confio can make extensions to CosmWasm to be used on multiple chains
